### PR TITLE
[TASK] Introduce Multi-Level TocTrees 

### LINKED ISSUE #197: toctree with 2 levels does not show second level
will provide a test for that one as well

### PR DESCRIPTION
**Pull request by linawolf**
Toctrees display subpages of subpages if `maxdepth` is not set or greater then one. The number of levels to be displayed depends on `:maxdepth:`. Only the main document title is used if `:titlesonly:` is set.

resolves #197

### CODE REVIEW

**Comment by wouterj** [comment]
Fyi, I'll put this one on my list to review tomorrow.

**Review by jwage — APPROVED** [review]

File: lib/HTML/Renderers/TocNodeRenderer.php [hunk 1]
```diff
@@ -49,11 +54,43 @@ public function render(): string
             $url = $this->environment->relativeUrl($reference->getUrl());
 
             $titleLinkUtility->buildLevel($url, $reference->getTitles(), 1, $tocItems, $file);
+            if ($maxDepth !== 0 && $maxDepth <= 1) {
```
Comment (wouterj) [inline lib/HTML/Renderers/TocNodeRenderer.php]:
```suggestion
            if ($maxDepth === 1) {
```

(unless I'm missing something)

File: lib/HTML/Renderers/TocNodeRenderer.php [hunk 2]
```diff
@@ -49,11 +54,43 @@ public function render(): string
...
+                continue;
+            }
+
+            $childTocItems = $this->buildSubPages($file, $titleLinkUtility, 1, $maxDepth);
+            if (count($tocItems) <= 0) {
+                continue;
+            }
+
+            $tocItems[count($tocItems) - 1]['subpages'] = $childTocItems;
         }
```
Comment (wouterj) [inline lib/HTML/Renderers/TocNodeRenderer.php]:
```suggestion
            if (count($tocItems) <= 0) {
                continue;
            }

            $tocItems[array_key_last($tocItems)]['subpages'] = $childTocItems;
        }
```

File: lib/HTML/Renderers/TocNodeRenderer.php [hunk 3]
```diff
@@ -49,11 +54,43 @@ public function render(): string
...
+        if ($metaData !== null) {
+            foreach ($metaData->getChildDocuments() as $childDocument) {
+                $reference = $this->resolver->resolve($this->environment, 'doc', $childDocument->getFile());
+                $titleLinkUtility->buildLevel($childDocument->getUrl(), $reference->getTitles(), 1, $tocItems, $childDocument->getFile());
+                if ($maxDepth !== 0 && $maxDepth <= $currentLevel + 1) {
+                    continue;
+                }
+
+                $childTocItems                              = $this->buildSubPages($childDocument->getFile(), $titleLinkUtility, $currentLevel + 1, $maxDepth);
+                $tocItems[count($tocItems) - 1]['subpages'] = $childTocItems;
```
Comment (wouterj) [inline lib/HTML/Renderers/TocNodeRenderer.php]:
```suggestion
                $tocItems[array_key_last($tocItems)]['subpages'] = $childTocItems;
```

**Review by wouterj — APPROVED** [review]
Nice work! Just some very minor comments.

I must admit that I can't 100% follow all logic, but the changes in test look correct to me so the logic must also be correct :)
I haven't been able to test this PR on the Symfony docs yet, as #251 had quite a big impact on our docs-builder package. I'm working on fixing 0.6 compatibility, but for now we can only run 0.5.

**Comment by linawolf** [comment]
Thank you @wouterj I committed your suggestions.

**Comment by linawolf** [comment]
@greg0ire can we merge this PR?

**Comment by greg0ire** [comment]
Can and did! Thanks @linawolf !

**Pull request merged**